### ISSUE TITLE
Maps not coerced to objects for input variables

### ISSUE THREAD

**Issue by davecardwell** [hashicorp/terraform#24092]
### Terraform Version

```
Terraform v0.12.20
```

#### main.tf
```hcl
module "child" {
  source = "./child"
  foo    = map("bar", "baz")
}
```

#### child/child.tf
```hcl
variable "foo" {
  type = object({ bar = string })
}
```

### Debug Output

https://gist.github.com/davecardwell/94c238e4f12d7d5612668a3747a35ea1

### Expected Behavior

From [Conversion of Complex Types](https://www.terraform.io/docs/configuration/types.html#conversion-of-complex-types) in the docs I expected the `map` to be converted to an `object` and accepted as a valid value.

### Actual Behavior

```
Error: Invalid value for module argument

  on main.tf line 3, in module "child":
   3:   foo    = map("bar", "baz")

The given value is not suitable for child module variable "foo" defined at
child/child.tf:1,1-15: object required.
```

### Steps to Reproduce
1. `terraform init`
2. `terraform validate`

**Comment by jbardin** [comment]
Hi @davecardwell,

Thanks, it appears the conversion functions aren't handling this case correctly.

The good news is that if you're only dealing with literals or expressions, they will initially be typed as an object and pass through. The `map` function itself is deprecated and should not be used.

**Comment by davecardwell** [comment]
Thanks for the speedy response, @jbardin.

I used the `map` function in my example but my actual use case is passing the result of an [External Data Source](https://www.terraform.io/docs/providers/external/data_source.html) to a child module.

I can set the variable type as `map(string)` and it works fine, or I can choose the individual values from the data source to pass through in the `module {}` block, so there are workarounds.

I went digging and raised [the issue](https://github.com/zclconf/go-cty/issues/39) with`go-cty` so will see if they think it’s viable to support map-to-object conversions. I don’t mind if you want to close this issue or leave it open for tracking progress there.
